pycbc/strain/strain.py

```diff
@@ -1597,12 +1597,28 @@ def advance(self, blocksize, timeout=10):
             strain = gate_data(strain, [(strain.start_time, 0., self.autogating_pad)])
             self.taper_immediate_strain = False
 
+        # apply gating if needed
+        if self.autogating_threshold is not None:
+            # the + 0 is for making a copy
+            glitch_times = detect_loud_glitches(
+                    strain + 0., psd_duration=1., psd_stride=.5,
+                    threshold=self.autogating_threshold,
+                    cluster_window=self.autogating_cluster,
+                    low_freq_cutoff=self.highpass_frequency,
+                    high_freq_cutoff=self.sample_rate/2,
+                    corrupt_time=self.autogating_pad)
+            if len(glitch_times) > 0:
+                logging.info('Autogating at %s',
+                             ', '.join(['%.3f' % gt for gt in glitch_times]))
+                gate_params = [[gt, self.autogating_window, self.autogating_pad] \
+                               for gt in glitch_times]
+                strain = gate_data(strain, gate_params)
```

Comment:
I am a bit worried about boundary effects here. At this point, with the current configuration, the segment of strain is only 8 s long. So it is not unlikely that a gate will fall too close to the start or end of the segment, and the tapering of the gate will be "chopped off" by the boundary, effectively making the gate shorter and with an abrupt edge. We can easily solve half of this problem by moving the block starting with `if len(glitch_times) > 0:` past the line reading `self.strain.start_time += blocksize`, and applying `gate_data` to `self.strain` instead of `strain`. Since `self.strain` also contains a lot of past strain, any gate that intersects the start of `strain` will be applied correctly. I am still not sure what to do for the gates that intersect the *end* of `strain`, though. We would need to "remember the gate" and apply the missing part to the *beginning* of `strain` at the next segment. Does this make sense?

Reply:
There is also the lazy option to simply discard any gate that intersects the edges of `strain`, I guess…